install/operator/pkg/syndesis/upgrade/database_upgrade.go

```diff
@@ -0,0 +1,414 @@
+	// this marks the end of the upgrade
+	if err := u.awaitScale(upgradeDeploymentName, newDeploymentTracker()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *databaseUpgrade) rollback() (err error) {
+	if err := u.deleteUpgrade(); err != nil {
+		u.log.Error(err, "Unable to delete database upgrade Deployment during rollback")
+	}
+
+	return nil
+}
+
+func (u *databaseUpgrade) canRun() bool {
+	if len(u.syndesis.Spec.Components.Database.ExternalDbURL) > 0 {
+		return false
+	}
+
+	target, err := u.target()
+	if err != nil {
+		u.log.Error(err, "Unable to determine target PostgreSQL version")
+		return false
+	}
+
+	current, err := u.current()
+	if err != nil {
+		u.log.Error(err, "Unable to determine current PostgreSQL version")
+		return false
+	}
+
+	u.name = fmt.Sprintf("Database upgrade from %f to %f", current, target)
+
+	u.log.Info("Determined versions of PostgreSQL database", "target", target, "current", current)
+	return target > current
+}
+
+// Connects to the running version of the database and queries it's version
+func (u *databaseUpgrade) currentFromRunningDatabase() (float64, error) {
+	config, err := configuration.GetProperties(
+		u.context,
+		configuration.TemplateConfig,
+		u.clientTools,
+		u.syndesis)
+
+	if err != nil {
+		return 0.0, err
+	}
+
+	databaseConfiguration := config.Syndesis.Components.Database
+
+	return util.PostgreSQLVersionAt(databaseConfiguration.User,
+		config.Syndesis.Components.Database.Password,
+		databaseConfiguration.Name,
+		databaseConfiguration.URL)
+}
+
+// Patches the `syndesis-db` DeploymentConfig setting the number
+// of replicas to 0
+func (u *databaseUpgrade) scaleDownDatabase() error {
+	u.log.Info("Scaling down the database deployment", "deployment", "syndesis-db")
+	if err := u.client().Patch(u.context, &oappsv1.DeploymentConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "syndesis-db",
+			Namespace: u.namespace,
+		},
+	}, client.RawPatch(types.MergePatchType, []byte(`{"spec":{"replicas":0}}`))); err != nil {
+		return err
+	}
+
+	if err := u.awaitScale("syndesis-db", newDeploymentConfigTracker()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Deploys the `syndesis-db-upgrade` Deployment with the same `syndesis-db-data` volume
+// as the current `syndesis-db` DeploymentConfig. Specifying `POSTGRESQL_UPGRADE=copy`
+// instructs the startup scripts within the centos/postgresql image to run pg_upgrade
+// to migrate the data files
+func (u *databaseUpgrade) deployUpgrade() error {
+	config, err := configuration.GetProperties(
+		u.context, configuration.TemplateConfig,
+		u.clientTools,
+		u.syndesis)
+	if err != nil {
+		return err
+	}
+
+	one := int32(1)
+	memory, err := resource.ParseQuantity(config.Syndesis.Components.Database.Resources.Memory)
+	if err != nil {
+		return err
+	}
+
+	metadata := upgradeMetadata.DeepCopy()
+	metadata.SetNamespace(u.namespace)
+
+	return u.client().Create(u.context, &appsv1.Deployment{
```

Comment:
Isn't better to have this in `pkg/syndesis/generator` and parse it here?

Reply:
To say that I dislike everything about `pkg/syndesis/generator` would be an understatement. Using format-fragile templating for data structures that are more succinct and easier to manage in code to an end result of compressed binary blobs that can't be diffed and are often left stale or result in conflicts. I fail to see what benefit does it give us over creating structs in code.